File: app/src/androidTest/java/br/com/jonathanzanella/myexpenses/account/ShowAccountActivityTest.java
package br.com.jonathanzanella.myexpenses.account;

import android.content.Intent;
import android.support.test.filters.MediumTest;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;

import org.joda.time.DateTime;
import org.junit.After;
import org.junit.Before;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.card.Card;
import br.com.jonathanzanella.myexpenses.card.CardRepository;
import br.com.jonathanzanella.myexpenses.database.DatabaseHelper;
import br.com.jonathanzanella.myexpenses.database.RepositoryImpl;
import br.com.jonathanzanella.myexpenses.expense.Expense;
import br.com.jonathanzanella.myexpenses.expense.ExpenseRepository;
import br.com.jonathanzanella.myexpenses.helpers.ActivityLifecycleHelper;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper;
import br.com.jonathanzanella.myexpenses.helpers.builder.CardBuilder;
import br.com.jonathanzanella.myexpenses.helpers.builder.ExpenseBuilder;
import br.com.jonathanzanella.myexpenses.helpers.builder.ReceiptBuilder;
import br.com.jonathanzanella.myexpenses.helpers.builder.SourceBuilder;
import br.com.jonathanzanella.myexpenses.receipt.Receipt;
import br.com.jonathanzanella.myexpenses.receipt.ReceiptRepository;
import br.com.jonathanzanella.myexpenses.source.Source;
import br.com.jonathanzanella.myexpenses.source.SourceRepository;

import static android.support.test.InstrumentationRegistry.getInstrumentation;
import static android.support.test.InstrumentationRegistry.getTargetContext;
import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.isDescendantOfA;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static br.com.jonathanzanella.myexpenses.helpers.UIHelper.matchToolbarTitle;
import static junit.framework.Assert.assertTrue;
import static org.hamcrest.core.AllOf.allOf;

@RunWith(AndroidJUnit4.class)
@MediumTest
public class ShowAccountActivityTest {
	private static final int ACCOUNT_BALANCE = 115;
	private static final int EXPENSE_VALUE = 25;
	private static final int RECEIPT_INCOME = 35;
	@Rule
	public ActivityTestRule<ShowAccountActivity> activityTestRule = new ActivityTestRule<>(ShowAccountActivity.class, true, false);

	private Account account;
	private AccountRepository repository;
	private ExpenseRepository expenseRepository;
	private ReceiptRepository receiptRepository;
	private SourceRepository sourceRepository;
	private CardRepository cardRepository;

	@Before
	public void setUp() throws Exception {
		new DatabaseHelper(getTargetContext()).recreateTables();

		repository = new AccountRepository(new RepositoryImpl<Account>(getTargetContext()));
		receiptRepository = new ReceiptRepository(new RepositoryImpl<Receipt>(getTargetContext()));
		expenseRepository = new ExpenseRepository(new RepositoryImpl<Expense>(getTargetContext()));
		cardRepository = new CardRepository(new RepositoryImpl<Card>(getTargetContext()), expenseRepository);
		sourceRepository = new SourceRepository(new RepositoryImpl<Source>(getTargetContext()));

		account = new Account();
		account.setName("test");
		account.setBalance(ACCOUNT_BALANCE);
		account.setAccountToPayCreditCard(true);
		repository.save(account);
	}

	@After
	public void tearDown() throws Exception {
		ActivityLifecycleHelper.closeAllActivities(getInstrumentation());
	}

	@Test
	public void shows_account_correctly() throws Exception {
		launchActivity();

		final String editAccountTitle = getTargetContext().getString(R.string.account) + " " + account.getName();
		matchToolbarTitle(editAccountTitle);

		String balanceAsCurrency = CurrencyHelper.format(account.getBalance());
		onView(withId(R.id.act_show_account_name)).check(matches(withText(account.getName())));
		onView(withId(R.id.act_show_account_balance)).check(matches(withText(balanceAsCurrency)));
	}

	@Test
	public void show_credit_card_bill_in_account_show_activity() {
		Card card = new CardBuilder().account(account).build(repository);
		assertTrue(cardRepository.save(card).isValid());
		Expense expense = new ExpenseBuilder().chargeable(card).build();
		assertTrue(expenseRepository.save(expense).isValid());

		launchActivity();

		String billName = getTargetContext().getString(R.string.invoice) + " " + card.getName();
		String value = CurrencyHelper.format(expense.getAmount());

		onView(withId(R.id.act_show_account_name)).check(matches(withText(account.getName())));
		onView(withId(R.id.name)).check(matches(withText(billName)));
		onView(withId(R.id.value)).check(matches(withText(value)));
	}

	@Test
	public void calculate_account_balance_correctly() throws Exception {
		generateTwoMonthsExpenses();
		generateTwoMonthsReceipts();

		launchActivity();

		int expectedBalance = ACCOUNT_BALANCE + RECEIPT_INCOME - EXPENSE_VALUE;
		String expectedValue = CurrencyHelper.format(expectedBalance);
		onView(allOf(
				withId(R.id.balance),
				isDescendantOfA(withId(R.id.thisMonth))))
				.check(matches(withText(expectedValue)));

		expectedBalance = expectedBalance + RECEIPT_INCOME - EXPENSE_VALUE;
		expectedValue = CurrencyHelper.format(expectedBalance);
		onView(allOf(
				withId(R.id.balance),
				isDescendantOfA(withId(R.id.nextMonth))))
				.check(matches(withText(expectedValue)));
	}

	private void launchActivity() {
		Intent i = new Intent();
		i.putExtra(ShowAccountActivity.Companion.getKEY_ACCOUNT_UUID(), account.getUuid());
		activityTestRule.launchActivity(i);
	}

	private void generateTwoMonthsReceipts() {
		Source s = new SourceBuilder().build();
		sourceRepository.save(s);

		Receipt receipt = new ReceiptBuilder()
				.income(RECEIPT_INCOME)
				.date(DateTime.now())
				.account(account)
				.source(s)
				.build();
		assertTrue(receiptRepository.save(receipt).isValid());
		receipt = new ReceiptBuilder()
				.income(RECEIPT_INCOME)
				.date(DateTime.now().plusMonths(1))
				.account(account)
				.source(s)
				.build();
		assertTrue(receiptRepository.save(receipt).isValid());
	}

	private void generateTwoMonthsExpenses() {
		Expense expense = new ExpenseBuilder()
				.value(EXPENSE_VALUE)
				.date(DateTime.now())
				.chargeable(account)
				.build();
		assertTrue(expenseRepository.save(expense).isValid());
		expense = new ExpenseBuilder()
				.value(EXPENSE_VALUE)
				.date(DateTime.now().plusMonths(1))
				.chargeable(account)
				.build();
		assertTrue(expenseRepository.save(expense).isValid());
	}
}

File: app/src/androidTest/java/br/com/jonathanzanella/myexpenses/account/ShowAccountActivityTest.kt
package br.com.jonathanzanella.myexpenses.account

import android.content.Intent
import android.support.test.InstrumentationRegistry.getInstrumentation
import android.support.test.InstrumentationRegistry.getTargetContext
import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.assertion.ViewAssertions.matches
import android.support.test.espresso.matcher.ViewMatchers
import android.support.test.filters.MediumTest
import android.support.test.rule.ActivityTestRule
import android.support.test.runner.AndroidJUnit4
import android.view.View
import br.com.jonathanzanella.TestApp
import br.com.jonathanzanella.myexpenses.App
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.card.CardDataSource
import br.com.jonathanzanella.myexpenses.expense.ExpenseDataSource
import br.com.jonathanzanella.myexpenses.helpers.ActivityLifecycleHelper
import br.com.jonathanzanella.myexpenses.helpers.UIHelper.matchToolbarTitle
import br.com.jonathanzanella.myexpenses.helpers.builder.CardBuilder
import br.com.jonathanzanella.myexpenses.helpers.builder.ExpenseBuilder
import br.com.jonathanzanella.myexpenses.helpers.builder.ReceiptBuilder
import br.com.jonathanzanella.myexpenses.helpers.builder.SourceBuilder
import br.com.jonathanzanella.myexpenses.helpers.toCurrencyFormatted
import br.com.jonathanzanella.myexpenses.receipt.ReceiptDataSource
import br.com.jonathanzanella.myexpenses.source.SourceDataSource
import com.facebook.testing.screenshot.Screenshot
import org.hamcrest.core.AllOf.allOf
import org.joda.time.DateTime
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import javax.inject.Inject

@RunWith(AndroidJUnit4::class)
@MediumTest
class ShowAccountActivityTest {
    @Rule @JvmField
    var activityTestRule = ActivityTestRule(ShowAccountActivity::class.java, true, false)

    private var account: Account? = null
    @Inject
    lateinit var dataSource: AccountDataSource
    @Inject
    lateinit var expenseDataSource: ExpenseDataSource
    @Inject
    lateinit var receiptDataSource: ReceiptDataSource
    @Inject
    lateinit var sourceDataSource: SourceDataSource
    @Inject
    lateinit var cardDataSource: CardDataSource

    @Before
    @Throws(Exception::class)
    fun setUp() {
        TestApp.getTestComponent().inject(this)

        App.resetDatabase()

        account = Account()
        account!!.name = "test"
        account!!.balance = ACCOUNT_BALANCE
        account!!.accountToPayCreditCard = true
        dataSource.save(account!!)
    }

    @After
    @Throws(Exception::class)
    fun tearDown() {
        ActivityLifecycleHelper.closeAllActivities(getInstrumentation())
    }

    @Test
    @Throws(Exception::class)
    fun shows_account_correctly() {
        launchActivity()

        val editAccountTitle = getTargetContext().getString(R.string.account) + " " + account!!.name
        matchToolbarTitle(editAccountTitle)

        val balanceAsCurrency = account!!.balance.toCurrencyFormatted()
        onView(ViewMatchers.withId(R.id.act_show_account_name)).check(matches(ViewMatchers.withText(account!!.name)))
        onView(ViewMatchers.withId(R.id.act_show_account_balance)).check(matches(ViewMatchers.withText(balanceAsCurrency)))

        Screenshot.snapActivity(activityTestRule.activity).record()
    }

    @Test
    @Throws(InterruptedException::class)
    fun show_credit_card_bill_in_account_show_activity() {
        val card = CardBuilder().account(account).build(dataSource)
        assert(cardDataSource.save(card).isValid)
        val expense = ExpenseBuilder().chargeable(card).build()
        assert(expenseDataSource.save(expense).isValid)

        launchActivity()

        val billName = getTargetContext().getString(R.string.invoice) + " " + card.name
        val value = expense.amount.toCurrencyFormatted()

        Thread.sleep(500)
        onView(ViewMatchers.withId(R.id.act_show_account_name)).check(matches(ViewMatchers.withText(account!!.name)))
        onView(ViewMatchers.withId(R.id.name)).check(matches(ViewMatchers.withText(billName)))
        onView(ViewMatchers.withId(R.id.value)).check(matches(ViewMatchers.withText(value)))

        Screenshot.snapActivity(activityTestRule.activity).record()
    }

    @Test
    @Throws(Exception::class)
    fun calculate_account_balance_correctly() {
        generateTwoMonthsExpenses()
        generateTwoMonthsReceipts()

        launchActivity()

        var expectedBalance = ACCOUNT_BALANCE + RECEIPT_INCOME - EXPENSE_VALUE
        var expectedValue = expectedBalance.toCurrencyFormatted()
        Thread.sleep(500)
        onView(allOf<View>(
                ViewMatchers.withId(R.id.balance),
                ViewMatchers.isDescendantOfA(ViewMatchers.withId(R.id.thisMonth))))
                .check(matches(ViewMatchers.withText(expectedValue)))

        expectedBalance = expectedBalance + RECEIPT_INCOME - EXPENSE_VALUE
        expectedValue = expectedBalance.toCurrencyFormatted()

        Thread.sleep(500)
        onView(allOf<View>(
                ViewMatchers.withId(R.id.balance),
                ViewMatchers.isDescendantOfA(ViewMatchers.withId(R.id.nextMonth))))
                .check(matches(ViewMatchers.withText(expectedValue)))

        Screenshot.snapActivity(activityTestRule.activity).record()
    }

    private fun launchActivity() {
        val i = Intent()
        i.putExtra(ShowAccountActivity.KEY_ACCOUNT_UUID, account!!.uuid)
        activityTestRule.launchActivity(i)
    }

    private fun generateTwoMonthsReceipts() {
        val s = SourceBuilder().build()
        sourceDataSource.save(s)

        var receipt = ReceiptBuilder()
                .income(RECEIPT_INCOME)
                .date(DateTime.now())
                .account(account)
                .source(s)
                .build()
        assert(receiptDataSource.save(receipt).isValid)
        receipt = ReceiptBuilder()
                .income(RECEIPT_INCOME)
                .date(DateTime.now().plusMonths(1))
                .account(account)
                .source(s)
                .build()
        assert(receiptDataSource.save(receipt).isValid)
    }

    private fun generateTwoMonthsExpenses() {
        var expense = ExpenseBuilder()
                .value(EXPENSE_VALUE)
                .date(DateTime.now())
                .chargeable(account)
                .build()
        assert(expenseDataSource.save(expense).isValid)
        expense = ExpenseBuilder()
                .value(EXPENSE_VALUE)
                .date(DateTime.now().plusMonths(1))
                .chargeable(account)
                .build()
        assert(expenseDataSource.save(expense).isValid)
    }

    companion object {
        private val ACCOUNT_BALANCE = 115
        private val EXPENSE_VALUE = 25
        private val RECEIPT_INCOME = 35
    }
}